File: smartcampost_mobile/lib/models/parcel_message.dart
class ParcelMessage {
  final String id;
  final String? parcelId;
  final String? senderAccountId;
  final String? senderRole;
  final String? senderName;
  final String content;
  final String? createdAt;
  final bool mine;
  final bool read;

  ParcelMessage({
    required this.id,
    this.parcelId,
    this.senderAccountId,
    this.senderRole,
    this.senderName,
    required this.content,
    this.createdAt,
    this.mine = false,
    this.read = false,
  });

  factory ParcelMessage.fromJson(Map<String, dynamic> json) => ParcelMessage(
    id: json['id']?.toString() ?? '',
    parcelId: json['parcelId']?.toString(),
    senderAccountId: json['senderAccountId']?.toString(),
    senderRole: json['senderRole'] as String?,
    senderName: json['senderName'] as String?,
    content: json['content'] as String? ?? '',
    createdAt: json['createdAt'] as String?,
    mine: json['mine'] as bool? ?? false,
    read: json['read'] as bool? ?? false,
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    if (parcelId != null) 'parcelId': parcelId,
    if (senderAccountId != null) 'senderAccountId': senderAccountId,
    if (senderRole != null) 'senderRole': senderRole,
    if (senderName != null) 'senderName': senderName,
    'content': content,
    if (createdAt != null) 'createdAt': createdAt,
    'mine': mine,
    'read': read,
  };
}

File: backend/src/main/java/com/smartcampost/backend/model/ParcelMessage.java
package com.smartcampost.backend.model;

import com.smartcampost.backend.model.enums.UserRole;
import jakarta.persistence.*;
import lombok.*;

import java.time.Instant;
import java.util.UUID;

/**
 * A single chat message exchanged between a parcel's client and the
 * agent/courier currently handling it (coordination, e.g. "I'm outside").
 */
@Entity
@Table(name = "parcel_message")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class ParcelMessage {

    @Id
    @Column(name = "message_id", columnDefinition = "BINARY(16)")
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parcel_id", nullable = false)
    private Parcel parcel;

    @Column(name = "sender_account_id", nullable = false, columnDefinition = "BINARY(16)")
    private UUID senderAccountId;

    @Enumerated(EnumType.STRING)
    @Column(name = "sender_role", nullable = false, length = 20)
    private UserRole senderRole;

    @Column(name = "sender_name", length = 150)
    private String senderName;

    @Lob
    @Column(name = "content", nullable = false, columnDefinition = "TEXT")
    private String content;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Builder.Default
    @Column(name = "read_by_recipient", nullable = false)
    private boolean readByRecipient = false;

    @PrePersist
    void onCreate() {
        if (id == null) id = UUID.randomUUID();
        if (createdAt == null) createdAt = Instant.now();
    }
}
